### ISSUE TITLE
CUDA OOM after 2000+ steps of training on Wan2.2

### ISSUE THREAD

**Issue by LittleNyima** [kohya-ss/musubi-tuner#450]
The training process for the Wan2.2 model runs successfully for an extended period but consistently fails with a CUDA Out of Memory (OOM) error after approximately 2048 steps. The issue is not an initial OOM, as the training starts and proceeds normally. I believe that this is caused by CUDA memory leakage but I have no idea about the cause.

**Environment**

```
torch version .................... 2.6.0+cu124
torch cuda version ............... 12.4
nvcc version ..................... 12.4
shared memory (/dev/shm) size .... 950.00 GB
GPU .............................. NVIDIA H100 80GB HBM3 x8
```

**Training command**

```
accelerate launch \
    --num_cpu_threads_per_process 1 \
    --mixed_precision fp16 \
    --num_processes ${ngpus} \
    --num_machines 1 \
    --machine_rank 0 \
    --main_process_ip 127.0.0.1 \
    --main_process_port 29500 \
    src/musubi_tuner/wan_train_network.py \
        --task i2v-A14B \
        --dit_high_noise weights/wan2.2_i2v_high_noise_14B_fp16.safetensors \
        --dit weights/wan2.2_i2v_low_noise_14B_fp16.safetensors \
        --offload_inactive_dit \
        --sdpa \
        --mixed_precision fp16 \
        --optimizer_type adamw \
        --learning_rate 2e-4 \
        --gradient_checkpointing \
        --max_data_loader_n_workers 2 \
        --persistent_data_loader_workers \
        --network_module networks.lora_wan \
        --network_dim 128 \
        --network_alpha 64 \
        --timestep_sampling shift \
        --discrete_flow_shift 5.0 \
        --max_train_epochs $max_train_epochs \
        --save_every_n_epochs 1 \
        --seed 42 \
        --output_dir $output_dir \
        --output_name wan2.2-i2v-a14b-high-low-lora \
        --dataset_config $dataset_config
```

**Exception traceback**

Please kindly refer to the screenshot below.

<img width="2902" height="1646" alt="Image" src="https://github.com/user-attachments/assets/9ceaefca-5ff5-41e0-b5dc-3ddf7187b3d6" />

**Comment by kohya-ss** [comment]
It certainly seems likely that there is a memory leak.
Does the amount of memory used increase gradually during training?

Also, if you know, please let me know what happens when training on a single GPU.

**Comment by LittleNyima** [comment]
> It certainly seems likely that there is a memory leak. Does the amount of memory used increase gradually during training?
> 
> Also, if you know, please let me know what happens when training on a single GPU.

I didn't monitor memory usage during the training. I will start a single-GPU training experiment after adding memory profiling, and I will report the results afterwards.

**Comment by Auryg** [comment]
Possibly related, using a 5090 on windows, on the training I ran overnight I noticed that the training slowed down over time.  It started at 3.8 s/it and by the time I woke up was down to 5.6 s/it.

accelerate launch --num_cpu_threads_per_process 1 --mixed_precision bf16 src/musubi_tuner/qwen_image_train_network.py --dit Q:/AI/Models/DiffusionModels/qwen_image_bf16.safetensors --vae Q:/AI/Models/VAE/qwen_vae_for_training.safetensors --text_encoder Q:/AI/Models/CLIP/qwen_2.5_vl_7b.safetensors --dataset_config S:/AI/Musubi/datasetTartoml --sdpa --mixed_precision bf16 --fp8_base --fp8_scaled --blocks_to_swap 10 --timestep_sampling qwen_shift --weighting_scheme none  --optimizer_type adamw8bit --learning_rate 2.5e-4 --lr_scheduler linear  --lr_scheduler_min_lr_ratio="7e-5" --gradient_checkpointing --max_data_loader_n_workers 2 --persistent_data_loader_workers --network_module networks.lora_qwen_image --network_dim 32 --max_train_epochs 20 --save_every_n_steps 250 --seed 42 

I have Nvidia's GPU memory offloading off for the system so *presumably* that wasn't happening, but if there is a memory leak that could explain it.

**Comment by LittleNyima** [comment]
I started a single-gpu training job, which ran for about 24 hours without any OOM issues. During training, its VRAM usage is around 72 GiB most of the time. However, at certain moments, its peak VRAM usage reaches 78GB, which may be the cause of OOM when training on multiple GPUs:

<img width="2004" height="1174" alt="Image" src="https://github.com/user-attachments/assets/35476ec1-a20e-4e99-8de0-77b0897692d4" />

For reference, the detailed status metrics of the GPU during training can be viewed from this link: [https://wandb.ai/littlenyima-ai/wan2.2-i2v-a14b-1gpu-train/runs/70wxs27g](https://wandb.ai/littlenyima-ai/wan2.2-i2v-a14b-1gpu-train/runs/70wxs27g)

**Comment by kohya-ss** [comment]
Thank you for sharing this information. There doesn't seem to be a memory leak, but there seem to be temporary high memory usage.
The peaks in memory usage may indicate the amount of memory actually needed, such as backward pass. I think it would be a good idea to lower the batch size a little to allow for a little headroom in memory usage.

I'm concerned that not only memory usage but also GPU usage is fluctuating.
It may be that data loading is not keeping up with the GPU speed. Could you try increasing `--max_data_loader_n_workers 2`, for example to 8 or 16?

**Comment by LittleNyima** [comment]
> Thank you for sharing this information. There doesn't seem to be a memory leak, but there seem to be temporary high memory usage. The peaks in memory usage may indicate the amount of memory actually needed, such as backward pass. I think it would be a good idea to lower the batch size a little to allow for a little headroom in memory usage.
> 
> I'm concerned that not only memory usage but also GPU usage is fluctuating. It may be that data loading is not keeping up with the GPU speed. Could you try increasing `--max_data_loader_n_workers 2`, for example to 8 or 16?

That makes sense. I agree with your assessment of the temporary memory peaks (and maybe I should offload several DiT blocks when training with 8 GPUs). And thanks for your advice about GPU usage. I'll launch a new experiment with `--max_data_loader_n_workers` increased to 16.

**Comment by kohya-ss** [comment]
I hope the new training works well. Note that block swap may not work with multi-GPU training.

**Comment by LittleNyima** [comment]
I conduct two experiments as follows:

1. Training with 8 GPUs with fp8_base precision, which works fine. (Though there is some visible quality loss compared to fp16).
2. Training with single GPU and increase `--max_data_loader_n_workers`, and the GPU state machine utilization is plotted in the figure below. (Red for `--max_data_loader_n_workers=16` and blue for `--max_data_loader_n_workers=2`) According to the figure, the increment in number of dataloader workers does not improve the GPU utilization, so I think dataloader is not the performance bottelneck.

<img width="2008" height="1150" alt="Image" src="https://github.com/user-attachments/assets/fb9a31b7-f8d3-4b15-861e-7ea2158df8c2" />

**Comment by LittleNyima** [comment]
> I conduct two experiments as follows:
> 
> 1. Training with 8 GPUs with fp8_base precision, which works fine. (Though there is some visible quality loss compared to fp16).
> 2. Training with single GPU and increase `--max_data_loader_n_workers`, and the GPU state machine utilization is plotted in the figure below. (Red for `--max_data_loader_n_workers=16` and blue for `--max_data_loader_n_workers=2`) According to the figure, the increment in number of dataloader workers does not improve the GPU utilization, so I think dataloader is not the performance bottelneck.

Some additional information: Since there were no other tasks scheduled on this server during the weekend, the CPU and memory usage were lower. After Monday (where the curve began to decline), the server load increased. I think this may be one of the reasons for the lower GPU utilization.

From this perspective, if we only observe the first half of the curve, increasing the number of workers does help improve GPU utilization.

**Comment by kohya-ss** [comment]
Thanks for the update.

The impact of max_data_loader_n_workers wasn't as great as I thought it would be. However, if the server is also being used for other tasks, I think 70-80% GPU usage isn't a bad number.

> * Training with 8 GPUs with fp8_base precision, which works fine. (Though there is some visible quality loss compared to fp16).

 `--fp8_scaled` option will improve the quality some extent with a minimal VRAM increase.

**Comment by LittleNyima** [comment]
Thank you very much for the update and the helpful suggestions.

Since we now have a stable working configuration and a reasonable explanation for the GPU usage, I consider the original issue resolved. I'll go ahead and close this issue.

Thanks again for all your help!
